lang/funcs/collection.go

```diff
@@ -497,7 +497,21 @@ var LookupFunc = function.New(&function.Spec{
 			return cty.NilType, fmt.Errorf("lookup() takes two or three arguments, got %d", len(args))
 		}
 
-		return args[0].Type().ElementType(), nil
+		ty := args[0].Type()
+		key := args[1].AsString()
+		switch {
+		case ty.IsObjectType():
+			if ty.HasAttribute(key) {
+				return args[0].GetAttr(key).Type(), nil
+			} else if len(args) == 3 {
+				// if the key isn't found but a default is provided,
+				// return the default type
+				return args[2].Type(), nil
+			}
+			return cty.DynamicPseudoType, nil
+		default:
+			return ty.ElementType(), nil
```

Comment:
Now that this argument is specified as `cty.DynamicPseudoType`, we're no longer guaranteed that it'll be a map here, so I think this needs an extra case for `ty.IsMapType()` and then for `default` to be an error saying that the given argument is not a map, to restore the type-checking behavior that was formerly done by the functions system automatically.

(Doing that in here guarantees that `Impl` will only be called if it's an object or a map, so there's no need to re-handle that problem in there.)